File: python/test/t_BoxCoxTransform_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott

ot.TESTPREAMBLE()


# Realization issued from a SpectralProcess
dimension = 1

# Parameters of the distribution
N = 101
t0 = 0.0
dt = 0.1
myTimeGrid = ot.RegularGrid(t0, dt, N)

# Create a Sample
# parameters of gaussien impose a few risk to get negative values
mySample = ot.Normal(10, 3).getSample(N)

# get a realization from distribution
myRealization = ot.TimeSeries(myTimeGrid, mySample)

# Create the lambda parameter
lambdaVector = ot.Point(dimension)
for index in range(dimension):
    lambdaVector[index] = (index + 2.0) * 0.1

myBoxCox = ot.BoxCoxTransform(lambdaVector)

print("myBoxCox=", myBoxCox)

# Get the input and output dimension
print("myBoxCox input dimension = ", myBoxCox.getInputDimension())
print("myBoxCox output dimension = ", myBoxCox.getOutputDimension())

# Evaluation of the BoxCoxTransform on the realization
print("input time series =")
print(myRealization)
print("output time series = ")
print(myBoxCox(myRealization))

print("gradient=", myBoxCox.gradient([0.5]))
print("hessian=", myBoxCox.hessian([0.5]))

# Call the getInverse method
myInverseBoxCox = myBoxCox.getInverse()
print("myInverseBoxCox = ", myInverseBoxCox)

# Get the number of calls
print("number of call(s) : ", myBoxCox.getCallsNumber())

# BoxCoxGradient direct coverage (no prints)
bc = ot.BoxCoxTransform([0.0, 1e-10, 0.5, 1.0])
_ = repr(bc.getGradient())
_ = str(bc.getGradient())
assert "BoxCoxGradient" in repr(bc.getGradient())
g = bc.gradient([1.0, 1.0, 4.0, 2.0])
ott.assert_almost_equal(g[0, 0], 1.0, 1e-8, 1e-8)
ott.assert_almost_equal(g[0, 3], 1.0, 1e-8, 1e-8)
with ott.assert_raises(Exception):
    bc.gradient([1.0])
with ott.assert_raises(Exception):
    bc.gradient([-5.0, 1.0, 1.0, 1.0])
_ = bc.getInverse().gradient([0.5, 0.1, 0.2, 0.3])
study = ot.Study()
study.setStorageManager(ot.XMLStorageManager("boxcox_tr.xml"))
study.add("t", bc)
study.save()

File: lib/test/t_BoxCoxTransform_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class BoxCoxTransform for standard methods
 *
 *  Copyright 2005-2022 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
  {
    // Realization issued from a Normal distribution
    const UnsignedInteger dimension = 1;

    // Fix the realization as a Normal
    // Parameters are done such as the values are positive
    Normal myDistribution(10, 3);

    // get a realization
    const UnsignedInteger n = 101;
    const Scalar timeStart = 0.;
    const Scalar timeStep = 0.1;
    Sample mySample(myDistribution.getSample(n));
    RegularGrid myTimeGrid(timeStart, timeStep, n);

    /* Create a TimeSeries */
    const TimeSeries myRealization(myTimeGrid, mySample);

    // Create the lambda parameter
    Point lambda(dimension);
    for (UnsignedInteger index = 0 ; index < dimension; ++index)
    {
      lambda[index] = (index + 2) * 0.1;
    }

    BoxCoxTransform myBoxCox(lambda);

    fullprint << "myBoxCox=" << myBoxCox << std::endl;

    /* Get the input and output dimension */
    fullprint << "myBoxCox input dimension=" << myBoxCox.getInputDimension() << std::endl;
    fullprint << "myBoxCox output dimension=" << myBoxCox.getOutputDimension() << std::endl;

    // BoxCoxTransform
    fullprint << "input time series=" << myRealization << std::endl;
    fullprint << "output time series=" << myBoxCox(myRealization) << std::endl;

    /* BoxCoxTransform's inverse accessor */
    InverseBoxCoxTransform myInverseBoxCox(myBoxCox.getInverse());
    fullprint << "myInverseBoxCox=" << myInverseBoxCox << std::endl;

    /* Get the number of calls */
    fullprint << "number of call(s) : " << myBoxCox.getCallsNumber() << std::endl;
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}
